aws-s3-transfer-manager/src/operation/upload_objects.rs

```diff
@@ -37,7 +37,15 @@ impl UploadObjects {
     ) -> Result<UploadObjectsHandle, crate::error::Error> {
         //  validate existence of source and return error if it's not a directory
         let source = input.source().expect("source set");
-        validate_target_is_dir(source).await?;
+        let metadata = tokio::fs::symlink_metadata(source).await?;
+        validate_target_is_dir(source).await.and_then(|_| {
+            if !input.follow_symlinks() && metadata.is_symlink() {
```

Comment:
Personally, I see this as a wrinkle in the type system brought by [`Metadata`](https://doc.rust-lang.org/std/fs/struct.Metadata.html), because we cannot tell how `Metadata` was originally obtained. When `fs::metadata` is used, the returned `Metadata` can only provide information about whether a path is a file or a directory, but it cannot tell if the path is a symlink. On the other hand, `fs::symlink_metadata` (which [corresponds to `lstat` on Unix](https://doc.rust-lang.org/std/fs/fn.symlink_metadata.html)) can identify whether a path is a symlink. However, if the path is a symlink, the `Metadata` returned by `fs::symlink_metadata` can only indicate that it is a symlink via `is_symlink()`. It will return false for both `is_file()` and `is_dir()`, even though the symlink may point to a file or directory.

> Can just use call metadata.is_dir() since we already looked up the metadata

Yes we can. For the reason above, we have to obtain `Metadata` twice in the worst case but for the default case where the upload operation does not follow symbolic links, we should obtain it only once (I overlooked this place as part commit 08330ab).

Reply:
Personally, I see this as a wrinkle in the type system brought by [`Metadata`](https://doc.rust-lang.org/std/fs/struct.Metadata.html), because we cannot tell how `Metadata` was originally obtained (but maybe by design). When `fs::metadata` is used, the returned `Metadata` can only provide information about whether a path is a file or a directory, but it cannot tell if the path is a symlink. On the other hand, `fs::symlink_metadata` (which [corresponds to `lstat` on Unix](https://doc.rust-lang.org/std/fs/fn.symlink_metadata.html)) can identify whether a path is a symlink. However, if the path is a symlink, the `Metadata` returned by `fs::symlink_metadata` can only indicate that it is a symlink via `is_symlink()`. It will return false for both `is_file()` and `is_dir()`, even though the symlink may point to a file or directory.

> Can just use call metadata.is_dir() since we already looked up the metadata

Yes we can. For the reason above, we have to obtain `Metadata` twice in the worst case but for the default case where the upload operation does not follow symbolic links, we should obtain it only once (I overlooked this place as part of commit 08330ab).

Reply:
e17439f obtains `Metadata` only once for the common case during validation.